surfer/viz.py

```diff
@@ -2682,19 +2717,67 @@ def _min_diff(self, beg, end):
             dr = np.array(end_d['r']) - np.array(beg_d['r'])
         return (np.array(dv), dr)
 
+    def _add_scalar_data(self, data):
+        # Add scalar values to dataset
+        array_id = self._mesh_dataset.point_data.add_array(data)
+        self._mesh_dataset.point_data.get_array(array_id).name = array_id
+        self._mesh_dataset.point_data.update()
+
+        # build visualization pipeline
+        with warnings.catch_warnings(record=True):
+            pipe = mlab.pipeline.set_active_attribute(
+                self._mesh_dataset, point_scalars=array_id, figure=self._f)
+            # The new data-source is added to the wrong figure by default
+            # (a Mayavi bug??)
+            self._f.add_child(pipe.parent)
+        self._mesh_clones[array_id] = pipe.parent
+        return array_id, pipe
+
+    def _remove_scalar_data(self, array_id):
+        "Removes scalar data"
+        self._mesh_clones.pop(array_id).remove()
+        self._mesh_dataset.point_data.remove_array(array_id)
+
     def add_overlay(self, old):
         """Add an overlay to the overlay dict from a file or array"""
-        surf = OverlayDisplay(old, figure=self._f)
-        for bar in ["pos_bar", "neg_bar"]:
-            try:
-                self._format_cbar_text(getattr(surf, bar))
-            except AttributeError:
-                pass
-        return surf
+        array_id, mesh = self._add_scalar_data(old.mlab_data)
+
+        with mayavi_threshold_patch:
```

Comment:
The two `threshold` calls need to be patched, I moved the context up because that way it has to be entered only once. Do you prefer to patch the two calls separately?